package/base-files/files/lib/upgrade/common.sh

```diff
@@ -231,11 +231,17 @@ indicate_upgrade() {
 # $(1): path to image
 # $(2): (optional) pipe command to extract firmware, e.g. dd bs=n skip=m
 default_do_upgrade() {
+	local err
 	sync
 	if [ -n "$UPGRADE_BACKUP" ]; then
 		get_image "$1" "$2" | mtd $MTD_ARGS $MTD_CONFIG_ARGS -j "$UPGRADE_BACKUP" write - "${PART_NAME:-image}"
 	else
 		get_image "$1" "$2" | mtd $MTD_ARGS write - "${PART_NAME:-image}"
 	fi
-	[ $? -ne 0 ] && exit 1
+	err=$?
+	if [ $err -ne 0 ] ; then
+		>&2 echo "ERROR: default_do_upgrade mtd write error: $err"
+		exit 1
+	fi
+	return $err
```

Comment:
This commit doesn't make sense:
On success, this function returns 0
On error, the entire sysupgrade procedure terminates, and there's no chance for calling function to deal with the error. I'll drop this commit and remove error checking in glinet.sh.

Reply:
> On success, this function returns 0

The existing code does _not_ return 0 on success. 

If the return status of `mtd` is 0, then the return status of `[ $? -ne 0 ]` is non-zero. The execution of `[` sets the return status to the result of the `test`. As a result, it is _not_ the return status of the `mtd` execution that is returned by the function.
```
root@OpenWrt:~# return_value_demo () {
> echo "This command succeeds"
> [ $? -ne 0 ] && echo "This would be an error exit"
> }
root@OpenWrt:~# return_value_demo && echo "succeeded" || echo "failed"
This command succeeds
failed
```

Not making this change seems to be a case of covering up existing faults in the existing code, rather than resolving them.

Further, it is challenging enough for end users to determine if a flash was successful while connected with serial. This at least provides a visible indication of the error condition before blindly rebooting into what might be a non-bootable condition, or one where no changes were made to the flashed firmware.

Reply:
> The existing code does not return 0 on success.

I'm referring to your patched version :)

> Further, it is challenging enough for end users to determine if a flash was successful while connected with serial. This at least provides a visible indication of the error condition before blindly rebooting into what might be a non-bootable condition, or one where no changes were made to the flashed firmware.

mtd itself will spit out error message to console if it fails. The added "exit 1" in previous commit is to solve confusing output like when mtd says it failed and sysupgrade script claims it succeeded.

> Not making this change seems to be a case of covering up existing faults in the existing code, rather than resolving them.

But this commit doesn't change any situation at all. Checking its return value only make sense when that value can report either a success or a failure. This function can only return on success and can't report any error to the calling script, so its return value just doesn't matter.